source/s3_client.c

```diff
@@ -1528,9 +1528,51 @@ static void s_s3_client_process_work_default(struct aws_s3_client *client) {
             AWS_LS_S3_CLIENT, "id=%p Updating connections, assigning requests where possible.", (void *)client);
         aws_s3_client_update_connections_threaded(client);
     }
+    /*******************/
+    /* Step 4: Cleanup Endpoints if Required. */
+    /*******************/
+
+    /* BEGIN CRITICAL SECTION */
+    aws_s3_client_lock_synced_data(client);
+
+    if (client->synced_data.process_endpoint_lifecycle_changes) {
+        client->synced_data.process_endpoint_lifecycle_changes = false;
+        for (struct aws_hash_iter iter = aws_hash_iter_begin(&client->synced_data.endpoints);
+             !aws_hash_iter_done(&iter);
+             aws_hash_iter_next(&iter)) {
+            struct aws_s3_endpoint *endpoint = (struct aws_s3_endpoint *)iter.element.value;
```

Comment:
this seems a bit scary: this can create 10s of things that need to happen on event loop. we also try to avoid locking this client too much from this thread as it can block other threads from doing their work. 

do we actually need every endpoint to have a separate termination event? can we simplify the logic and have a simple event that repeats every n seconds and cleans up whatever endpoints do not have work? (i dont think we care if endoint gets cleaned up in 3s exactly or worst case in 6s). we do something similar for dns from what i remember

Reply:
Thanks, Created a new https://github.com/awslabs/aws-c-s3/pull/419 with the new logic.